File: app/src/main/java/com/example/golpezero/MainActivity.java
package com.example.golpezero;

import static android.content.Context.MODE_PRIVATE;
import static androidx.activity.result.ActivityResultCallerKt.registerForActivityResult;
import static androidx.core.content.ContextCompat.registerReceiver;

import android.content.BroadcastReceiver;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.core.content.ContextCompat;

public class MainActivity {

    private static final String TAG = "PermissaoApp";
    private static final String PREFS_NAME = "AppPrefs";
    private static final String PERMISSIONS_GRANTED_KEY = "PermissionsGranted";

    private BroadcastReceiver callReceiver;
    private Button btnAtivar;
    private TextView tvStatus;

    // Launcher que recebe o resultado da solicitação
    private final ActivityResultLauncher<String[]> requestPermissionsLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestMultiplePermissions(), permissions -> {
                // Verificamos se as permissões essenciais foram concedidas
                boolean allGranted = permissions.getOrDefault(android.Manifest.permission.READ_PHONE_STATE, false) &&
                        permissions.getOrDefault(android.Manifest.permission.READ_CALL_LOG, false);

                if (allGranted) {
                    Log.i(TAG, "Todas as permissões essenciais foram concedidas!");
                    ativarMonitoramento();
                } else {
                    Log.w(TAG, "Permissões necessárias foram negadas.");
                    Toast.makeText(this, "A proteção não pode ser ativada sem as permissões.", Toast.LENGTH_LONG).show();
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        btnAtivar = findViewById(R.id.btn_ativar_protecao);
        tvStatus = findViewById(R.id.tv_status_protecao);

        // LÓGICA DE PERSISTÊNCIA: Verificar no início se as permissões já foram dadas antes.
        if (verificarSePermissoesJaConcedidas()) {
            ativarMonitoramento();
        } else {
            // Prepara a UI para solicitar permissão
            tvStatus.setText("Proteção Inativa");
            btnAtivar.setText("Ativar Proteção Agora");
            btnAtivar.setOnClickListener(v -> verificarEsolicitarPermissoes());
        }
    }

    private void verificarEsolicitarPermissoes() {
        // Lista de permissões que nosso app precisa para a funcionalidade principal
        String[] permissoesNecessarias = {
                android.Manifest.permission.READ_PHONE_STATE,
                android.Manifest.permission.READ_CALL_LOG
        };

        // Lança a solicitação para as permissões que ainda não foram concedidas
        requestPermissionsLauncher.launch(permissoesNecessarias);
    }

    /**
     * ESTA É A FUNÇÃO MAIS IMPORTANTE PÓS-PERMISSÃO
     */
    private void ativarMonitoramento() {
        Log.d(TAG, "Ativando o monitoramento de chamadas...");

        // 1. REGISTRAR O BROADCASTRECEIVER:
        // Agora que temos permissão, podemos registrar nosso "ouvido" para escutar as chamadas do sistema.
        // Sem permissão, o registro falharia ou o app quebraria.
        callReceiver = new CallReceiver();
        IntentFilter filter = new IntentFilter(android.telephony.TelephonyManager.ACTION_PHONE_STATE_CHANGED);
        registerReceiver(callReceiver, filter);

        // 2. ATUALIZAR A INTERFACE DO USUÁRIO (UI):
        // Fornecer feedback visual de que a proteção está ativa.
        tvStatus.setText("Proteção Ativa ");
        btnAtivar.setText("Monitoramento Ativado");
        btnAtivar.setEnabled(false);


        SharedPreferences.Editor editor = getSharedPreferences(PREFS_NAME, MODE_PRIVATE).edit();
        editor.putBoolean(PERMISSIONS_GRANTED_KEY, true);
        editor.apply();

        Toast.makeText(this, "Monitoramento de chamadas ativado!", Toast.LENGTH_SHORT).show();
    }

    private boolean verificarSePermissoesJaConcedidas() {
        SharedPreferences prefs = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);
        // Também verificamos a permissão real, caso o usuário a tenha revogado nas configurações do Android.
        boolean permissionsPreviouslyGranted = prefs.getBoolean(PERMISSIONS_GRANTED_KEY, false);
        boolean readPhoneStateGranted = ContextCompat.checkSelfPermission(this, android.Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED;

        return permissionsPreviouslyGranted && readPhoneStateGranted;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();

        if (callReceiver != null) {
            unregisterReceiver(callReceiver);
        }
    }

}

File: app/src/main/java/com/example/golpezero/KK/MainActivity.kt
package com.example.golpezero

import android.Manifest
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Bundle
import android.telephony.TelephonyManager
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.content.ContextCompat.registerReceiver

class MainActivity : AppCompatActivity() {

    companion object {
        private const val TAG = "CallMonitor"
        private const val PREFS_NAME = "CallMonitorPrefs"
        private const val KEY_IS_MONITORING = "permissions_granted"
    }

    private lateinit var btnToggleMonitor: Button
    private lateinit var tvStatus: TextView
    private var callReceiver: CallReceiver? = null

    // CORRETO: registerForActivityResult já é fornecido pela Activity
    private val requestPermissionsLauncher = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        val allGranted = permissions.all { it.value }

        if (allGranted) {
            enableMonitoring()
        } else {
            Toast.makeText(this, "Permissões negadas.", Toast.LENGTH_SHORT).show()
            updateUI(false)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(com.example.golpezero.R.layout.activity_main)

        btnToggleMonitor = findViewById(com.example.golpezero.R.id.btnToggleMonitor)
        tvStatus = findViewById(com.example.golpezero.R.id.tvStatus)

        btnToggleMonitor.setOnClickListener { toggleMonitoring() }
        checkInitialState()
    }

    private fun checkInitialState() {
        if (hasRequiredPermissions()) {
            enableMonitoring()
        } else {
            updateUI(false)
        }
    }

    private fun toggleMonitoring() {
        if (hasRequiredPermissions()) {
            if (callReceiver == null) {
                enableMonitoring()
            } else {
                disableMonitoring()
            }
        } else {
            requestRequiredPermissions()
        }
    }

    private fun hasRequiredPermissions(): Boolean {
        return ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.READ_PHONE_STATE
        ) == PackageManager.PERMISSION_GRANTED && ContextCompat.checkSelfPermission(
            this,
            Manifest.permission.READ_CALL_LOG
        ) == PackageManager.PERMISSION_GRANTED
    }

    private fun requestRequiredPermissions() {
        val permissionsToRequest = arrayOf(
            Manifest.permission.READ_PHONE_STATE,
            Manifest.permission.READ_CALL_LOG
        ).filter {
            ContextCompat.checkSelfPermission(this, it) != PackageManager.PERMISSION_GRANTED
        }.toTypedArray()

        if (permissionsToRequest.isNotEmpty()) {
            requestPermissionsLauncher.launch(permissionsToRequest)
        } else {
            enableMonitoring()
        }
    }

    private fun enableMonitoring() {
        callReceiver = CallReceiver()
        val filter = IntentFilter(TelephonyManager.ACTION_PHONE_STATE_CHANGED)
        registerReceiver(callReceiver, filter)

        updateUI(true)
        Toast.makeText(this, "Monitoramento ativado", Toast.LENGTH_SHORT).show()
    }

    private fun disableMonitoring() {
        callReceiver?.let {
            unregisterReceiver(it)
            callReceiver = null
        }

        updateUI(false)
        Toast.makeText(this, "Monitoramento desativado", Toast.LENGTH_SHORT).show()
    }

    private fun updateUI(isMonitoring: Boolean) {
        if (isMonitoring) {
            tvStatus.text = "Monitoramento ATIVO"
            btnToggleMonitor.text = "Desativar"
        } else {
            tvStatus.text = "Monitoramento INATIVO"
            btnToggleMonitor.text = "Ativar"
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        callReceiver?.let {
            unregisterReceiver(it)
        }
    }
}

private fun MainActivity.unregisterReceiver(it: CallReceiver) {
    TODO("Not yet implemented")
}
